### PR TITLE
OHRM5X-65: Develop Qualification Skill Screen

### PR DESCRIPTION
**Pull request by KaveeshaSilva1996**

### CODE REVIEW

**Comment by KaveeshaSilva1996** [comment]
I will fix the conflicts and the refactoring issues

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    /**
+     * @inheritDoc
+     * @throws DaoException
+     * @throws Exception
+     */
+    public function delete(): EndpointDeleteResult
+    {
+        // TODO:: Check data group permission
+        $ids = $this->getRequestParams()->getArray(RequestParams::PARAM_TYPE_BODY, CommonParams::PARAMETER_IDS);
+        $this->getSkillService()->deleteSkills($ids);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:172]: Did you test with empty array?

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    public function saveSkill(): Skill
+    {
+        // TODO:: Check data group permission
+        $id = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, CommonParams::PARAMETER_ID);
+        $name = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_NAME);
+        $description = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_DESCRIPTION);
+        if (!empty($id)) {
+            $skill = $this->getSkillService()->getSkillById($id);
+            if ($skill == null) {
+                throw new RecordNotFoundException('No Record Found');
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:190]: Remove custom message just throw new RecordNotFoundException();

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        $skill->setName($name);
+        $skill->setDescription($description);
+        return $this->getSkillService()->saveSkill($skill);
+    }
+
+    /**
+     * @return Response
+     * @throws DaoException
+     */
+    public function deleteSkills(): Response
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:205]: do we need this method?

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 4]
```diff
@@ -0,0 +1,189 @@
+<?php
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:2]: remove this blank line

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+/**
+ * OrangeHRM is a comprehensive Human Resource Management (HRM) System that captures
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:4]: use this code header on other files as well

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+ */
+
+namespace OrangeHRM\Admin\Dao;
+
+use Doctrine\ORM\QueryBuilder;
+use Doctrine\ORM\Tools\Pagination\Paginator;
+use OrangeHRM\Admin\Dto\SkillSearchFilterParams;
+use OrangeHRM\Entity\Skill;
+use OrangeHRM\ORM\Doctrine;
+use \DaoException;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:28]: `use OrangeHRM\Core\Exception\DaoException;`

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+use \Exception;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:29]: `use Exception;`

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    public function getSkillById(int $id): ?Skill
+    {
+        try {
+            return Doctrine::getEntityManager()->getRepository(Skill::class)->find($id);
+        } catch (Exception $e) {
+            throw new DaoException($e->getMessage());
+        }
+    }
+
+    public function getSkillByName($name)
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:63]: this method not converted.

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+            $q = Doctrine::getEntityManager()->createQueryBuilder();
+            $q->delete(Skill::class, 's')
+                ->where($q->expr()->in('s.id', $toDeleteIds));
+            return $q->getQuery()->execute();
+        } catch (Exception $e) {
+            throw new DaoException($e->getMessage());
+        }
+    }
+
+    public function isExistingSkillName($skillName)
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:93]: this also not converted, remove if not using

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    private function _buildSearchQuery(SkillSearchFilterParams $skillSearchParams): QueryBuilder
+    {
+        $q = Doctrine::getEntityManager()->getRepository(
+            Skill::class
+        )->createQueryBuilder('s');
+
+        if (!is_null($skillSearchParams->getSortField())) {
+            $q->addOrderBy($skillSearchParams->getSortField(), $skillSearchParams->getSortOrder());
+        }
+        if (!empty($skillSearchParams->getLimit())) {
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:139]: careful when using empty with integer, will return true for 0

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+     *
+     * @param SkillSearchFilterParams $skillSearchParams
+     * @return int
+     * @throws DaoException
+     */
+    public function getSearchSkillsCount(SkillSearchFilterParams $skillSearchParams): int
+    {
+        try {
+            $q = $this->_buildSearchQuery($skillSearchParams);
+            $paginator = new \OrangeHRM\ORM\Paginator($q);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:183]: import this as `use OrangeHRM\ORM\Paginator`

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+     * @return Skill
+     * @throws DaoException
+     * @throws RecordNotFoundException
+     */
+    public function saveSkill(): Skill
+    {
+        // TODO:: Check data group permission
+        $id = $this->getRequestParams()->getInt(RequestParams::PARAM_TYPE_ATTRIBUTE, CommonParams::PARAMETER_ID);
+        $name = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_NAME);
+        $description = $this->getRequestParams()->getString(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_DESCRIPTION);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:186]: is this correct? isn't description optinal?
```php
$description = $this->getRequestParams()->getStringOrNull(RequestParams::PARAM_TYPE_BODY, self::PARAMETER_DESCRIPTION);
```

File: symfony/plugins/orangehrmAdminPlugin/Service/SkillService.php [hunk 13]
```diff
@@ -0,0 +1,143 @@
...
+    }
+
+    /**
+     * Checks whether the given skill name exists
+     *
+     * Case insensitive
+     *
+     * @param string $skillName Skill name that needs to be checked
+     * @return boolean
+     * @version 2.6.12
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Service/SkillService.php:106]: remove this @version

File: symfony/plugins/orangehrmAdminPlugin/Service/SkillService.php [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+     */
+    public function getSearchSkillsCount(SkillSearchFilterParams $skillSearchParams): int
+    {
+        try {
+            return $this->getSkillDao()->getSearchSkillsCount($skillSearchParams);
+        } catch (Exception $e) {
+            throw new ServiceException($e->getMessage(), $e->getCode(), $e);
+        }
+    }
+
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Service/SkillService.php:140]: remove blank lines

File: symfony/plugins/orangehrmAdminPlugin/config/routes.yaml [hunk 15]
```diff
@@ -40,6 +40,34 @@ apiv2_admin_users:
   defaults:
     _api: OrangeHRM\Admin\Api\UserAPI
 
+apiv2_admin_employment_status:
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/config/routes.yaml:43]: why `apiv2_admin_employment_status` for skills. rebase the branch

File: symfony/plugins/orangehrmAdminPlugin/entity/Skill.php [hunk 16]
```diff
@@ -19,38 +21,82 @@ class Skill
...
-    private $id;
+    private int $id;
 
     /**
      * @var string
      *
      * @ORM\Column(name="name", type="string", length=120)
      */
-    private $name;
+    private string $name;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/entity/Skill.php:31]: `name`, `description` both fields are nullable, check the ohrm_skill table

File: symfony/plugins/orangehrmAdminPlugin/test/Dao/SkillDaoTest.php [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+use OrangeHRM\Admin\Dao\SkillDao;
+use OrangeHRM\Admin\Dto\SkillSearchFilterParams;
+use OrangeHRM\Config\Config;
+use OrangeHRM\Core\Api\V2\RequestParams;
+use OrangeHRM\Entity\Skill;
+use OrangeHRM\Tests\Util\TestCase;
+use OrangeHRM\Tests\Util\TestDataService;
+
+/**
+ * @group Admin
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Dao/SkillDaoTest.php:31]: add @group Dao as well
```php
/**
 * @group Admin
 * @group Dao
```

File: symfony/plugins/orangehrmAdminPlugin/test/Dao/SkillDaoTest.php [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+        $result = $this->skillDao->deleteSkills($toBedeletedIds);
+        $this->assertEquals($result, 2);
+    }
+
+    public function testSearchSkill(): void
+    {
+        $skillSearchParams = new SkillSearchFilterParams();
+
+        $result = $this->skillDao->searchSkill($skillSearchParams);
+        $this->assertEquals(3,count($result) );
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Dao/SkillDaoTest.php:74]: format the code using IDE, may have to format all files

File: symfony/plugins/orangehrmAdminPlugin/test/Service/SkillServiceTest.php [hunk 19]
```diff
@@ -0,0 +1,130 @@
...
+use OrangeHRM\Admin\Dao\SkillDao;
+use OrangeHRM\Admin\Dto\SkillSearchFilterParams;
+use OrangeHRM\Admin\Service\SkillService;
+use OrangeHRM\Config\Config;
+use OrangeHRM\Entity\Skill;
+use OrangeHRM\Tests\Util\TestCase;
+use OrangeHRM\Tests\Util\TestDataService;
+
+/**
+ * @group Admin
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Service/SkillServiceTest.php:31]: add @group Service

File: symfony/plugins/orangehrmAdminPlugin/test/Service/SkillServiceTest.php [hunk 20]
```diff
@@ -0,0 +1,130 @@
...
+    private $skillService;
+    protected $fixture;
+
+    /**
+     * Set up method
+     */
+    protected function setUp(): void
+    {
+        $this->skillService = new SkillService();
+        $this->fixture = Config::get('sf_plugins_dir') . '/orangehrmAdminPlugin/test/fixtures/SkillDao.yml';
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Service/SkillServiceTest.php:45]: ohrm_plugins_dir

File: symfony/plugins/orangehrmAdminPlugin/test/Service/SkillServiceTest.php [hunk 21]
```diff
@@ -0,0 +1,130 @@
...
+        TestDataService::populate($this->fixture);
+    }
+
+    public function testDeleteSkill(): void
+    {
+        $toBeDeletedSkillIds = array(1, 2);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Service/SkillServiceTest.php:51]: refer [this](https://stackoverflow.com/questions/23697248/phpstorm-replace-array-notation-to-short-syntax) to convert array() to []

File: symfony/plugins/orangehrmPimPlugin/entity/EmployeeSkill.php [hunk 22]
```diff
@@ -24,11 +26,12 @@ class EmployeeSkill
     /**
      * @var int
      *
-     * @ORM\Column(name="skill_id", type="integer")
-     * @ORM\Id
-     * @ORM\GeneratedValue(strategy="NONE")
+     * @ORM\ManyToOne(targetEntity="OrangeHRM\Entity\Skill", mappedBy="employeeSkills")
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmPimPlugin/entity/EmployeeSkill.php:29]: refactor this entity as well. Since this also related to skills.

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

**Comment by samanthajayasinghe** [comment]
@KaveeshaSilva1996 can you rebase the branch ?

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+ * if not, write to the Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA
+ *
+ */
+
+namespace OrangeHRM\Admin\Api;
+
+use OrangeHRM\Admin\Dto\SkillSearchFilterParams;
+use OrangeHRM\Core\Api\CommonParams;
+use \OrangeHRM\Core\Exception\SearchParamException;
```
[the remark this answers is not in the document]
Comment (KaveeshaSilva1996) [reply symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:26]: fixed

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php (hunk 3 above)
Comment (KaveeshaSilva1996) [reply symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:205]: removed

File: symfony/plugins/orangehrmPimPlugin/entity/EmployeeSkill.php (hunk 22 above)
Comment (KaveeshaSilva1996) [reply symfony/plugins/orangehrmPimPlugin/entity/EmployeeSkill.php]: Refactored this relationship

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+use OrangeHRM\Core\Api\V2\Serializer\EndpointGetAllResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointGetOneResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointUpdateResult;
+use OrangeHRM\Entity\Skill;
+use OrangeHRM\Admin\Service\SkillService;
+use OrangeHRM\Admin\Api\Model\SkillModel;
+use Orangehrm\Rest\Api\Exception\RecordNotFoundException;
+use Orangehrm\Rest\Http\Response;
+use OrangeHRM\Core\Exception\DaoException;
+use \Exception;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:46]: `use Exception`

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+use OrangeHRM\Core\Api\V2\Endpoint;
+use OrangeHRM\Core\Exception\ServiceException;
+
+class SkillAPI extends Endpoint implements CrudEndpoint
+{
+    const PARAMETER_NAME = 'name';
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:52]: `public const PARAMETER_NAME = 'name';`

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php [hunk 26]
```diff
@@ -0,0 +1,268 @@
...
+use OrangeHRM\Core\Api\V2\RequestParams;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointCreateResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointDeleteResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointGetAllResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointGetOneResult;
+use OrangeHRM\Core\Api\V2\Serializer\EndpointUpdateResult;
+use OrangeHRM\Entity\Skill;
+use OrangeHRM\Admin\Service\SkillService;
+use OrangeHRM\Admin\Api\Model\SkillModel;
+use Orangehrm\Rest\Api\Exception\RecordNotFoundException;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:43]: this is wrong.
`use OrangeHRM\Core\Api\V2\Exception\RecordNotFoundException;`

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php [hunk 27]
```diff
@@ -0,0 +1,268 @@
...
+use Orangehrm\Rest\Http\Response;
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:44]: why `use Orangehrm\Rest\Http\Response;`

File: symfony/plugins/orangehrmAdminPlugin/Dao/EmploymentStatusDao.php [hunk 28]
```diff
@@ -98,9 +99,9 @@ public function searchEmploymentStatus(EmploymentStatusSearchFilterParams $emplo
 
     /**
      * @param EmploymentStatusSearchFilterParams $employmentStatusSearchParams
-     * @return QueryBuilder
+     * @return Paginator
      */
-    private function _buildSearchQuery(EmploymentStatusSearchFilterParams $employmentStatusSearchParams): QueryBuilder
+    private function getSearchEmploymentStatusPaginator(EmploymentStatusSearchFilterParams $employmentStatusSearchParams): \OrangeHRM\ORM\Paginator
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/EmploymentStatusDao.php:104]: format the code with `Optimize imports`

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 29]
```diff
@@ -0,0 +1,160 @@
...
+{
+    /**
+     * @param Skill $skill
+     * @return Skill
+     * @throws DaoException
+     */
+    public function saveSkill(Skill $skill): Skill
+    {
+        try {
+            Doctrine::getEntityManager()->persist($skill);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:41]: Update to `$this->persist(...)`

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 30]
```diff
@@ -0,0 +1,160 @@
...
+
+    /**
+     * @param int $id
+     * @return object|null|Skill
+     * @throws DaoException
+     */
+    public function getSkillById(int $id): ?Skill
+    {
+        try {
+            return Doctrine::getEntityManager()->getRepository(Skill::class)->find($id);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:57]: update to `$this->getRepository(...)`

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 31]
```diff
@@ -0,0 +1,160 @@
...
+
+    /**
+     * @param array $toDeleteIds
+     * @return int
+     * @throws DaoException
+     */
+    public function deleteSkills(array $toDeleteIds): int
+    {
+        try {
+            $q = Doctrine::getEntityManager()->createQueryBuilder();
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:71]: update to `$this->createQueryBuilder(....)`

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 32]
```diff
@@ -0,0 +1,160 @@
...
+     * Search Skill
+     *
+     * @param SkillSearchFilterParams $skillSearchParams
+     * @return array
+     * @throws DaoException
+     */
+    public function searchSkill(SkillSearchFilterParams $skillSearchParams): array
+    {
+        try {
+            $q = $this->getSearchSkillPaginator($skillSearchParams);
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:91]: `$q` => `$paginator`

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 33]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+
+    /**
+     * @param SkillSearchFilterParams $skillSearchParams
+     * @return Paginator
+     */
+    private function getSearchSkillPaginator(SkillSearchFilterParams $skillSearchParams): Paginator
+    {
+        $q = Doctrine::getEntityManager()->getRepository(
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:104]: update to `$this->createQueryBuilder(...)`

File: symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php [hunk 34]
```diff
@@ -0,0 +1,160 @@
...
+    /**
+     * Get Employment Statuses
+     *
+     * @return Skill[]
+     * @throws DaoException
+     */
+    public function getSkills(): array
+    {
+        try {
+            return Doctrine::getEntityManager()->getRepository(
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dao/SkillDao.php:136]: update to `$this->getRepository(...)`

File: symfony/plugins/orangehrmAdminPlugin/Dto/SkillSearchFilterParams.php [hunk 35]
```diff
@@ -0,0 +1,69 @@
...
+ * Boston, MA  02110-1301, USA
+ */
+
+namespace OrangeHRM\Admin\Dto;
+
+use OrangeHRM\Core\Dto\FilterParams;
+
+class SkillSearchFilterParams extends FilterParams
+{
+    public const ALLOWED_SORT_FIELDS = ['s.name','s.description'];
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/Dto/SkillSearchFilterParams.php:26]: format the code

File: symfony/plugins/orangehrmAdminPlugin/entity/Skill.php [hunk 36]
```diff
@@ -19,38 +21,83 @@ class Skill
...
-    private $EmployeeSkill;
+    private Collection $employeeSkills;
+
 
     /**
      * Constructor
      */
     public function __construct()
     {
-        $this->EmployeeSkill = new \Doctrine\Common\Collections\ArrayCollection();
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/entity/Skill.php:53]: better to keep this. read the doctrine documentation

File: symfony/plugins/orangehrmAdminPlugin/entity/Skill.php [hunk 37]
```diff
@@ -19,38 +21,83 @@ class Skill
...
      * @ORM\Column(name="name", type="string", length=120)
      */
-    private $name;
+    private ?string $name;
 
     /**
-     * @var string
+     * @var string | null
      *
      * @ORM\Column(name="description", type="string", length=2147483647)
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/entity/Skill.php:36]: remove length & type should be `text`. and add `nullable=true`

File: symfony/plugins/orangehrmAdminPlugin/entity/Skill.php [hunk 38]
```diff
@@ -19,38 +21,83 @@ class Skill
...
      * @ORM\GeneratedValue(strategy="AUTO")
      */
-    private $id;
+    private int $id;
 
     /**
-     * @var string
+     * @var string | null
      *
      * @ORM\Column(name="name", type="string", length=120)
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/entity/Skill.php:29]: `@ORM\Column(name="name", type="string", length=120, nullable=true)`

File: symfony/plugins/orangehrmAdminPlugin/test/Service/SkillServiceTest.php [hunk 39]
```diff
@@ -0,0 +1,133 @@
...
+    protected string $fixture;
+
+    /**
+     * Set up method
+     * @throws Exception
+     */
+    protected function setUp(): void
+    {
+        $this->skillService = new SkillService();
+        $this->fixture = Config::get('ohrm_plugins_dir') . '/orangehrmAdminPlugin/test/fixtures/SkillDao.yml';
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAdminPlugin/test/Service/SkillServiceTest.php:48]: `Config::get(Config::PLUGINS_DIR)`

File: symfony/plugins/orangehrmAuthenticationPlugin/config/routes.yaml [hunk 40]
```diff
@@ -6,7 +6,7 @@ auth_login:
 auth_validate:
   path: /auth/validate
   controller: OrangeHRM\Authentication\Controller\ValidateController::handle
-  methods: [ POST ]
+  methods: [ GET, POST ]
```
Comment (RajithaKumara) [inline symfony/plugins/orangehrmAuthenticationPlugin/config/routes.yaml:9]: revert this change

**Review by RajithaKumara — CHANGES_REQUESTED** [review]

File: dbscript/dbscript-2.sql [hunk 41]
```diff
@@ -238,7 +238,7 @@ INSERT INTO `hs_hr_country`(`cou_code`, `name`, `cou_name`, `iso3`, `numcode`) V
     ('ZM', 'ZAMBIA', 'Zambia', 'ZMB', 894),
```
[the remark this answers is not in the document]
Comment (KaveeshaSilva1996) [reply dbscript/dbscript-2.sql:238]: reverted

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/EditQualificationSkill.vue [hunk 42]
```diff
@@ -0,0 +1,170 @@
...
+            v-model="skill.description"
+            :rules="rules.description"
+          />
+        </oxd-form-row>
+
+        <oxd-divider />
+
+        <oxd-form-actions>
+          <oxd-button displayType="ghost" label="Cancel" @click="onCancel" />
+          <oxd-button
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/EditQualificationSkill.vue:52]: Please use `<submit-button />`

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/SaveQualificationSkill.vue [hunk 43]
```diff
@@ -0,0 +1,150 @@
...
+          return (v && v.length < 50) || 'Should be less than 50 characters';
+        });
+        this.rules.name.push(v => {
+          const index = data.findIndex(item => item.name == v);
+          return index === -1 || 'Skill name should be unique';
+        });
+        this.isLoading = false;
+      })
+      .catch(error => {
+        console.log(error);
```
Comment (Super-Chama) [inline symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/SaveQualificationSkill.vue:146]: please remove console logs. if any specific error you want to show, add a toast message. HTTP requests errors are anyway intercepted and displayed.

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/EditQualificationSkill.vue (hunk 42 above)
Comment (KaveeshaSilva1996) [reply symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/EditQualificationSkill.vue]: added

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/SaveQualificationSkill.vue (hunk 43 above)
Comment (KaveeshaSilva1996) [reply symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/SaveQualificationSkill.vue]: removed

File: symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php (hunk 27 above)
Comment (KaveeshaSilva1996) [reply symfony/plugins/orangehrmAdminPlugin/Api/SkillAPI.php:44]: removed

File: symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/EditQualificationSkill.vue [hunk 44]
```diff
@@ -0,0 +1,164 @@
...
+        // Fetch list data for unique test
+        return this.http.getAll();
+      })
+      .then(response => {
+        const {data} = response.data;
+        this.rules.name.push(v => {
+          return (!!v && v.trim() !== '') || 'Required';
+        });
+        this.rules.name.push(v => {
+          return (v && v.length <= 100) || 'Should be less than 50 characters';
```
Comment (samanthajayasinghe) [inline symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/EditQualificationSkill.vue:145]: Should be less than 100 characters

Comment (KaveeshaSilva1996) [reply symfony/client/src/orangehrmAdminPlugin/pages/qualificationSkill/EditQualificationSkill.vue]: Corrected this as "return (v && v.length <= 50) || 'Should be less than 50 characters';"
Since, as Chamara developed when saving, it allows only 50 characters.

**Review by samanthajayasinghe — APPROVED** [review]
